lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,778 @@
+    """
+
+    # The FileTree class owns the definition of the "controller" level
+    # directory where PUT will store uploading files. Co-locating this
+    # with the ARCHIVE tree ensures that we can move files without an
+    # additional copy, and the upload will already fail if the file
+    # system has insufficient space. Defining the directory here allows
+    # FileTree discovery to ignore it.
+    TEMPORARY = "UPLOAD"
+
+    def __init__(self, options: PbenchServerConfig, logger: Logger):
+        """
+        Construct a FileTree object. We don't do any discovery here, because
+        the mutation operations allow dynamic minimal discovery to save on
+        some time. The `full_discovery` method allows full discovery when
+        desired.
+
+        Args:
+            options: PbenchServerConfig configuration object
+            logger: A Pbench python Logger
+        """
+        self.options = options
+        self.archive_root = self.options.ARCHIVE
+        self.incoming_root = self.options.INCOMING
+        self.results_root = self.options.RESULTS
+        self.logger = logger
+        self.controllers: Dict[str, Controller] = {}
+        self.datasets: Dict[str, Tarball] = {}
+
+    def full_discovery(self):
+        """
+        Discover and diagnose the state of the entire file tree, including the
+        ARCHIVE, INCOMING, and RESULTS subtrees.
+
+        This is useful for standalone reporting, including to enable an audit
+        check. Generally it's overkill for specific operations such as adding
+        or removing a dataset. This setup is sufficient but not necessary for
+        mutation operations.
+        """
+        self._discover_filesystem()
+
+    def __contains__(self, dataset: str) -> bool:
+        """
+        Allow asking whether a FileTree contains an entry for a specific
+        dataset.
+
+        Args:
+            dataset: Dataset name
+
+        Returns:
+            True if the dataset is present
+        """
+        return dataset in self.datasets
+
+    def __getitem__(self, dataset: str) -> Dataset:
+        """
+        Direct access to a dataset Tarball object by name.
+
+        Args:
+            dataset: Dataset name
+
+        Returns:
+            Tarball object
+        """
+        try:
+            return self.datasets[dataset]
+        except KeyError:
+            raise DatasetNotFound(dataset)
+
+    def _clean_empties(self, controller: str):
+        """
+        Remove empty controller directories from the RESULTS and INCOMING
+        trees. If there are no remaining tarballs in the ARCHIVE controller
+        directory, remove all empty state subdirectories; and, if the
+        controller directory is now empty remove it as well.
+
+        Args:
+            controller: Name of the controller to clean up
+        """
+        results = self.options.RESULTS / controller
+        if results.exists() and not list(results.iterdir()):
```

Comment:
Maybe; I tried to keep everything pure `Path` rather than dropping out to `os` operations. Part of that was that I was hoping to unit test by mocking `Path` ... which turned out to be vastly messier than I actually was willing to attempt ... so the "purity" is a bit less of a priority now. Would `any(results.iterdir())` work as well?

Reply:
I think it would.  The internet's wisdom was that it was less efficient, but my own concern is avoiding producing the entire list which I think it would do.